File: 06 - Bridge/Dart/bridge.dart
void main() {
  RasterRendere raster = RasterRendere();
  VectorRenderer vector = VectorRenderer();
  Circle circle = Circle(raster, 2);
  circle.draw();
  circle.resize(5);
  circle.draw();
}

abstract class Renderer {
  void renderCircle(int radius);
}

class VectorRenderer implements Renderer {
  @override
  void renderCircle(int radius) {
    print('Drawing a circle of radius $radius');
  }
}

class RasterRendere implements Renderer {
  @override
  void renderCircle(int radius) {
    print('Drawing pixels for a circle of radius $radius');
  }
}

abstract class Shape {
  final Renderer renderer;
  Shape(this.renderer);

  void draw() {}
  void resize(int factor) {}
}

class Circle extends Shape {
  int radius;
  Renderer renderer;
  Circle(this.renderer, this.radius) : super(renderer);

  @override
  void draw() {
    renderer.renderCircle(radius);
  }

  @override
  void resize(int factor) {
    radius = radius * factor;
  }
}

File: 06 - Bridge/Python/bridge.py
from abc import ABC

class Renderer(ABC):
    def render_circle(self, radius):
        pass

class VectorRenderer(Renderer):
    def render_circle(self, radius):
        print(f'Drawing a circle of radius {radius}')
    
class RasterRendere(Renderer):
    def render_circle(self, radius):
        print(f'Drawing pixels for a circle of radius {radius}')

class Shape:
    def __init__(self, renderer) -> None:
        self.renderer = renderer
    
    def draw(self):
        pass
    
    def resize(self, factor):
        pass

class Circle(Shape):
    def __init__(self, renderer, radius) -> None:
        super().__init__(renderer)
        self.radius = radius
    
    def draw(self):
        self.renderer.render_circle(self.radius)
    
    def resize(self, factor):
        self.radius = self.radius * factor


if __name__ == "__main__":
    raster = RasterRendere()
    vector = VectorRenderer()
    circle = Circle(raster, 2)
    circle.draw()
    circle.resize(5)
    circle.draw()
